### PR TITLE
feat: update to TS 2.9 and enable strictFunctionTypes

### LINKED ISSUE #12177: Typescript version advancement to 2.9.x
#### Bug, feature request, or proposal:
Feature

#### What is the expected behavior?
Angular updated it's typescript dependency from 2.7.x to 2.9.x yesterday with 6.0.9 release.
Is Angular Material following anytime soon?

#### What is the current behavior?
Typescript dependency locked to 2.7.x

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Angular 6.0.9
Material 6.3.3

### PR DESCRIPTION
**Pull request by crisbeto**
Bumps the project to TypeScript 2.9, enables the `strictFunctionTypes` everywhere and fixes the various errors.

Fixes #12177.

### CODE REVIEW

File: src/cdk/layout/breakpoints-observer.ts [hunk 1]
```diff
@@ -75,16 +75,16 @@ export class BreakpointObserver implements OnDestroy {
...
     // Create callback for match changes and add it is as a listener.
-    const queryObservable = fromEventPattern(
+    const queryObservable = fromEventPattern<MediaQueryList>(
       // Listener callback methods are wrapped to be placed back in ngZone. Callbacks must be placed
       // back into the zone because matchMedia is only included in Zone.js by loading the
       // webapis-media-query.js file alongside the zone.js file.  Additionally, some browsers do not
       // have MediaQueryList inherit from EventTarget, which causes inconsistencies in how Zone.js
       // patches it.
-      (listener: MediaQueryListListener) => {
+      (listener: Function) => {
```
Comment (mmalerba) [inline src/cdk/layout/breakpoints-observer.ts]: Is there no type more specific than `Function` that we can use here?

Comment (crisbeto) [reply src/cdk/layout/breakpoints-observer.ts]: I don't think so. The `strictFunctionTypes` seem to resolve it to `Function`. A workaround it is to have `as MediaQueryListener` before calling, but that seemed a little ugly.

Comment (mmalerba) [reply src/cdk/layout/breakpoints-observer.ts]: Aright, sg

**Comment by josephperrott** [comment]
@crisbeto Is this still a WIP?  If not go ahead and update the title and mark it merge ready again.

**Comment by crisbeto** [comment]
It is still is, I haven't been able to track down the Bazel issue yet.

**Comment by crisbeto** [comment]
Looks like we're blocked here until https://github.com/angular/angular/issues/24521 is resolved. cc @jelbourn

**Comment by jelbourn** [comment]
@crisbeto getting this:
```
src/cdk/scrolling/virtual-scroll-viewport.ts:146:15 - error TS2345: Argument of type 'MonoTypeOperatorFunction<void>' is not assignable to parameter of type 'OperatorFunction<Event, void>'.
  Types of parameters 'source' and 'source' are incompatible.
    Type 'Observable<Event>' is not assignable to type 'Observable<void>'.
      Type 'Event' is not assignable to type 'void'.

146               startWith<void>(null!),

```

**Comment by jelbourn** [comment]
@crisbeto I looked at that error w/ Miles and it goes away if we change that line to:
```ts
startWith<Event | null>(null),
```

**Comment by crisbeto** [comment]
That's odd @jelbourn. I had fixed that error yesterday and I don't get any diff changes or compilation errors locally. Re-pushing to see if that fixes it.

**Comment by jelbourn** [comment]
Still failing on that line- it must be something different about Google's internal TS setup

**Comment by crisbeto** [comment]
Could be. I've changed it to `Event | null`.

**Pull request merged**